domain/modeldefaults/state/state.go

```diff
@@ -230,10 +235,276 @@ WHERE m.uuid = ?
+		}
+		return errors.Trace(err)
+	})
+	return cloud.UUID(cld.UUID), errors.Annotatef(err, "getting cloud UUID for %q", cloudName)
+}
+
+// UpdateCloudDefaults is responsible for updating default config values for a
+// cloud. This function will allow the addition and updating of attributes.
+// If the cloud doesn't exist, an error satisfying [clouderrors.NotFound]
+// is returned.
+func (st *State) UpdateCloudDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	updateAttrs map[string]string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cld := dbCloud{UUID: cloudUUID.String()}
+
+	upsertStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_defaults (cloud_uuid, key, value)
+VALUES ($cloudDefaultValue.*)
+ON CONFLICT(cloud_uuid, key) DO UPDATE
+    SET value = excluded.value
+    WHERE cloud_uuid = excluded.cloud_uuid
+    AND key = excluded.key;
+`, cloudDefaultValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for k, v := range updateAttrs {
+			err := tx.Query(ctx, upsertStmt, cloudDefaultValue{UUID: cld.UUID, Key: k, Value: v}).Run()
+			// The cloud UUID has previously been checked. This allows us to avoid having to use RunAtomic.
+			if database.IsErrConstraintForeignKey(err) {
+				return interrors.Errorf("cloud %q not found%w", cloudUUID, errors.Hide(clouderrors.NotFound))
+			} else if err != nil {
+				return interrors.Errorf("updating cloud default keys %q: %w", cloudUUID, err)
+			}
+		}
+		return nil
+	})
+}
+
+// DeleteCloudDefaults deletes the specified cloud default
+// config values if they exist.
+func (st *State) DeleteCloudDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	removeAttrs []string,
+) error {
+	if len(removeAttrs) == 0 {
+		return nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cld := dbCloud{UUID: cloudUUID.String()}
+	toRemove := attrs(removeAttrs)
+
+	deleteStmt, err := st.Prepare(`
+DELETE FROM cloud_defaults
+WHERE key IN ($attrs[:])
+AND cloud_uuid = $dbCloud.uuid;
+`, toRemove, cld)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, deleteStmt, toRemove, cld).Run()
+	})
+	if err != nil {
+		return interrors.Errorf("removing cloud default keys for %q: %w", cloudUUID, err)
+	}
+	return nil
+}
+
+// UpdateCloudRegionDefaults is responsible for updating default config values
+// for a cloud region. This function will allow the addition and updating of
+// attributes. If the cloud is not found an error satisfying [clouderrors.NotFound]
+// is returned. If the region is not found, am error satisfying [errors.NotFound]
+// is returned.
+func (st *State) UpdateCloudRegionDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	regionName string,
+	updateAttrs map[string]string,
+) error {
+	if len(updateAttrs) == 0 {
```

Comment:
This breaks the contract set out above.  My suggestion would be to remove this check all together.

Reply:
How? The contract says nothing about there needing to be > 0 attributes specified.
I'll move the check up to the service. We don't want to waste cycles if there's nothing to do.